singularity-eos/base/variadic_utils.hpp

```diff
@@ -16,13 +16,28 @@
 #define SINGULARITY_EOS_BASE_VARIADIC_UTILS_HPP_
 
 #include <type_traits>
+#include <utility>
 
 namespace singularity {
 namespace variadic_utils {
 
 // Some generic variatic utilities
 // ======================================================================
 
+// SFINAE to check if a value is a null ptr
+template <typename T, typename = typename std::enable_if<
+                          std::is_pointer<typename std::remove_reference<
+                              typename std::remove_cv<T>::type>::type>::value>::type>
+constexpr inline bool is_nullptr(T &&t) {
+  return std::forward<T>(t) == nullptr;
+}
+template <typename T,
+          typename std::enable_if<!std::is_pointer<typename std::remove_reference<
+              typename std::remove_cv<T>::type>::type>::value>::type * = nullptr>
+constexpr inline bool is_nullptr(T &&) {
+  return false;
+}
```

Comment:
This can be cleaned up when we move to C++17. Thanks to @dholladay00 for the particular version of this, which checks edge cases.

Reply:
It's too late on a Friday for understanding the nestedness going on with `::type>::type>::value>::type>` 🤯

Reply:
Haha blame/thank @dholladay00 for that one. Conceptually all it's doing is checking whether or not T is a pointer. If T is a pointer, check if it's null. Otherwise, return false.

All the other nested type nonsense is checking for edge cases, references to pointers, const pointers, etc.

Reply:
The C++17 version will need to retain the nested templates, but will be able to move it out of `std::enable_if` and into `if constexpr`

Reply:
This is something we did in EAP to handle this. It might make sense to implement this to just make things slightly more readable and reusable if we need to do this elsewhere:
```c++
// C++14 implementation of std::remove_cvref (availalbe since C++20)
template<typename T>
struct remove_cvref { typedef std::remove_cv_t<std::remove_reference_t<T>> type; };

// Helper types
template<typename T>
using remove_cvref_t = typename remove_cvref<T>::type;
```

Credit to Diego via CJ for our implementation

Reply:
:+1: Done.